[jit] fix tensor.data attribute to not requires_grad

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25350 [jit] fix tensor.data attribute to not requires_grad**

Summary:
This fix the .data attribute where it should not require_grad for its returned value and use whatever python use to get the data with version_counter=0 always

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25350

Differential Revision: [D17100829](https://our.internmc.facebook.com/intern/diff/D17100829)

wanchaol committed:
[jit] add set_grad_enabled to TorchScript and fix data attribute

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

wanchaol committed:
Update on "[jit] add set_grad_enabled to TorchScript and fix data attribute"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

wanchaol committed:
Update on "[jit] add set_grad_enabled to TorchScript and fix data attribute"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D17100829](https://our.internmc.facebook.com/intern/diff/D17100829)

wanchaol committed:
Update on "[jit] add set_grad_enabled to TorchScript and fix data attribute"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D17100829](https://our.internmc.facebook.com/intern/diff/D17100829)

zdevito commented on test/test_jit.py:
```diff
@@ -612,6 +612,22 @@ def forward(self, input):
+            # type: (Tensor) -> Tuple[Tensor, Tensor]
+            torch.set_grad_enabled(True)
+            y = x * 2
+            torch.set_grad_enabled(False)
```
This is going to leave gradients off after this test, right? That seems like it will break other tests.

wanchaol replied:
hmm yeah you are right, there's no API to get the current grad_mode so it's hard to restore the environment as before. We should actually enable the context manager version first instead of acting like a function. I will postpone adding this until we have context manager supported.

zdevito commented on test/test_jit.py:
```diff
@@ -612,6 +612,22 @@ def forward(self, input):
+            return y, z
+
+        x = torch.randn(2, 2, requires_grad=True)
+        print(test_set_grad_mode_function.graph)
```
stray print

zdevito commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -650,6 +655,13 @@ RegisterOperators reg(
          },
          aliasAnalysisFromSchema()),
+     Operator(
+         "aten::set_grad_enabled(bool mode) -> ()",
```
This needs to be marked as conservative. Otherwise the compiler can move compilation above/below this function during optimizations.

zdevito reviewed (commented):
A couple comments inline. Also, I would appreciate some context in the PR description. It is currently blank.

wanchaol committed:
Update on "[jit] fix tensor.data attribute to not requires_grad"


Summary:
This fix the .data attribute where it should not require_grad for its returned value and use whatever python use to get the data with version_counter=0 always

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25350

Differential Revision: [D17100829](https://our.internmc.facebook.com/intern/diff/D17100829)

wanchaol committed:
Update on "[jit] fix tensor.data attribute to not requires_grad"


Summary:
This fix the .data attribute where it should not require_grad for its returned value and use whatever python use to get the data with version_counter=0 always

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25350

Differential Revision: [D17100829](https://our.internmc.facebook.com/intern/diff/D17100829)
